File: examples/cpp/discover.cpp
// Author: Lars Op den Kamp (lars@opdenkamp-it.nl)
// Copyright (c) 2026 Op den Kamp IT Solutions

/*
 * Discovers MX Remote devices and prints what they report.
 *
 * The same program as examples/c/discover.c, through the C++ header: the
 * handle is released by leaving main, and the callbacks are virtual methods on
 * an object that already holds what they need.
 *
 * Usage: discover [interface-address]
 */

#include "mx_remote.hpp"

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <thread>

namespace {

std::atomic<bool> running(true);

void stop(int) { running = false; }

class Printer : public mxr::Handler {
public:
    /// The client the snapshots are read from. Set after it is created, which
    /// is in time because nothing calls back before it is started.
    mxr::Remote *remote = nullptr;

    void on_device_update(mxr::Uid device) override {
        mxr_device_info_t info;
        if (remote == nullptr || mxr_device(remote->get(), device, &info) != MXR_OK) return;
        std::printf("  %s  %-16s %-12s %-16s protocol 0x%02x, %zu bays%s\n",
                    device.str().c_str(), info.model, info.serial, info.name,
                    info.supported_protocol, info.bay_count, info.online ? "" : " (offline)");
    }

    void on_bay_update(mxr::BayUid bay) override {
        mxr_bay_info_t info;
        if (remote == nullptr || mxr_bay(remote->get(), bay, &info) != MXR_OK) return;
        std::printf("    bay %u %-16s %s\n", bay.port(), info.user_name,
                    info.signal_detected == MXR_TRUE ? "signal" : "no signal");
    }
};

} // namespace

int main(int argc, char **argv) {
    Printer printer;
    mxr_config_t config{};
    config.name = "discover";
    if (argc > 1) config.local_ip = argv[1];

    mxr::Remote remote = mxr::Remote::open(&config, &printer);
    if (!remote) {
        std::fprintf(stderr, "could not create the client: %s\n", mxr::last_error().c_str());
        return 1;
    }
    printer.remote = &remote;

    if (remote.start() != MXR_OK) {
        std::fprintf(stderr, "could not start: %s\n", mxr::last_error().c_str());
        return 1;
    }

    std::string ip;
    uint16_t port = 0;
    if (remote.target(ip, port))
        std::printf("listening, sending to %s:%u. Ctrl-C to stop.\n", ip.c_str(), port);

    std::signal(SIGINT, stop);
    std::signal(SIGTERM, stop);
    while (running) std::this_thread::sleep_for(std::chrono::seconds(1));

    std::vector<mxr::Uid> devices = remote.devices();
    std::printf("\n%zu device(s):\n", devices.size());
    for (std::vector<mxr::Uid>::const_iterator it = devices.begin(); it != devices.end(); ++it)
        printer.on_device_update(*it);

    /* remote closes and frees itself here, before printer goes out of scope. */
    return 0;
}

File: mx-remote/examples/discover.rs
// Author: Lars Op den Kamp (lars@opdenkamp-it.nl)
// Copyright (c) 2026 Op den Kamp IT Solutions

//! Discovers MX Remote devices and prints what they report.
//!
//! The shape here is the one most programs want: two handler methods that say
//! only which device or bay moved, and a snapshot read back for the detail.
//! Both run on the library's receive thread, so what they do is kept short.
//!
//! Usage: `cargo run --example discover [interface-address]`

use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::{Arc, OnceLock};
use std::time::Duration;

use mx_remote::{BayUid, Config, DeviceUid, EventHandler, Remote};

/// The client the snapshots are read from.
///
/// A handler is handed to the client that will call it, so it cannot hold one
/// at the time it is built. It is filled in before the client is started,
/// which is before anything can call back.
static CLIENT: OnceLock<Arc<Remote>> = OnceLock::new();

struct Printer;

impl EventHandler for Printer {
    fn on_device_update(&self, device: DeviceUid) {
        let Some(remote) = CLIENT.get() else { return };
        let Some(info) = remote.device(device) else {
            return;
        };
        println!(
            "  {device}  {:<16} {:<12} {:<16} protocol {:#04x}, {} bays{}",
            info.model,
            info.serial,
            info.name,
            info.supported_protocol,
            info.bays.len(),
            if info.online { "" } else { " (offline)" },
        );
    }

    fn on_bay_update(&self, bay: BayUid) {
        let Some(remote) = CLIENT.get() else { return };
        let Some(info) = remote.bay(bay) else { return };
        println!(
            "    bay {} {:<16} {}",
            bay.port,
            info.user_name,
            match info.signal_detected {
                Some(true) => "signal",
                _ => "no signal",
            },
        );
    }
}

fn main() -> std::io::Result<()> {
    let mut config = Config::default();
    config.name = Some("discover".to_owned());
    if let Some(address) = std::env::args().nth(1) {
        config.local_ip = Some(address.parse().map_err(|_| {
            std::io::Error::new(std::io::ErrorKind::InvalidInput, "not an IPv4 address")
        })?);
    }

    let remote = Arc::new(Remote::new(config, Arc::new(Printer))?);
    let _ = CLIENT.set(Arc::clone(&remote));
    remote.start()?;

    if let Some(target) = remote.target() {
        println!("listening, sending to {target}. Ctrl-C to stop.");
    }

    let running = Arc::new(AtomicBool::new(true));
    let stop = Arc::clone(&running);
    ctrl_c(move || stop.store(false, Ordering::Relaxed));
    while running.load(Ordering::Relaxed) {
        std::thread::sleep(Duration::from_millis(200));
    }

    let devices = remote.devices();
    println!("\n{} device(s):", devices.len());
    for device in devices {
        Printer.on_device_update(device);
    }

    remote.close();
    Ok(())
}

/// Runs `f` on the first Ctrl-C.
///
/// The library takes no signal handler of its own - it is meant to link into a
/// host program that has its own - so an example that wants one installs it.
fn ctrl_c(f: impl FnOnce() + Send + 'static) {
    static HANDLER: OnceLock<()> = OnceLock::new();
    let _ = HANDLER.set(());
    std::thread::spawn(move || {
        let mut line = String::new();
        // Reading to end of input is the portable stand-in for a signal
        // handler: Ctrl-C closes the terminal's read on most shells, and
        // Ctrl-D ends it everywhere.
        let _ = std::io::stdin().read_line(&mut line);
        f();
    });
}
